engine/scene/sceneloader.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "component.h"
+
+#include <nlohmann/json.hpp>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace pong
+{
+
+class GameObject;
+
+class SceneLoader
+{
+public:
+    SceneLoader();
+    ~SceneLoader() = default;
+
+    void PreLoadScenes();
+    std::vector<std::unique_ptr<GameObject>> LoadScene(const std::string& sceneName);
+
+private:
+    std::vector<std::unique_ptr<GameObject>> LoadSceneFromJson(const nlohmann::json& sceneJson);
+
+    std::unordered_map<std::string, nlohmann::json> mScenes {};
```

Comment:
One thing to note here is that you're gluing yourself to `nlohmann::json`.

Any time a tps shows up in a header, it means that there is a transitive physical dependency on this tps, and it "leaks" into the rest of your codebase.

e.g. if `A.h` has `json` dependency, `B.h` includes `A.h`, any `C.h` that includes `B.h` then needs to know and care about `json.


There are ways to get around this, whicih we can talk about, but it may not be worth it in the short run. Just trying to get you to think about dependencies and how they restrict future changes. Some of the best designed codebases are designed primarily to be Easy To Change as a primary intention.

Would changing your dependency to `nlohmann::json` be easy or hard? What would necessitate a change like that? Is it worth it to design around a possible change like that? (architectural questions to ask yourself)

Reply:
Yeah that is a good point. Now that I have worked with C++ a lot more, it seems like #include can create a lot of different annoying issues. Makes me want modules a lot more haha.

Reply:
Yeah modules would be great for this